sql: plumb a soft/hard flag for the limit hint

RaduBerinde opened the pull request:
There are situations where if we know for sure the upper layer only needs X rows
we can be more efficient: Nate is looking at making the sortNode use only O(X)
memory in this case. This change improves the SetLimitHint API by adding a
soft vs hard limit flag. Soft limits are "hints", hard limits are "promises".

@nvanbenschoten

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="35" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/5270)

<!-- Reviewable:end -->

nvanbenschoten commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>, thanks for taking care of this!

---

Reviewed 11 of 12 files at r1.
Review status: 11 of 12 files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---

*[sql/union.go, line 159 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5270#-KCvuqfkjif6i2mYHC6S:-KCvuqfkjif6i2mYHC6T:738871553) ([raw file](https://github.com/cockroachdb/cockroach/blob/d21582821e6360810f4b5778f82c61ea5fc8fcd7/sql/union.go#L159)):*
What's the reasoning behind this one defaulting both to hard?

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5270#-:-KCvv-9xPjCZxhHyP1AK:190606870)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
Review status: 11 of 12 files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---

*[sql/union.go, line 159 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5270#-KCvuqfkjif6i2mYHC6S:-KCvvleSOwMZW3GZ61vO:-20267110) ([raw file](https://github.com/cockroachdb/cockroach/blob/d21582821e6360810f4b5778f82c61ea5fc8fcd7/sql/union.go#L159)):*
We can probably do better here depending on the case (UNION, UNION DISTINCT, EXCEPT, INTERSECT), but I think in most cases it is possible to have to read more rows than we eventually emit.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5270)*
<!-- Sent from Reviewable.io -->

nvanbenschoten commented:
Review status: 11 of 12 files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---

*[sql/union.go, line 159 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5270#-KCvuqfkjif6i2mYHC6S:-KCvwSvPX_l2PMOUo6O9:67032775) ([raw file](https://github.com/cockroachdb/cockroach/blob/d21582821e6360810f4b5778f82c61ea5fc8fcd7/sql/union.go#L159)):*
Ah right, I read that as defaulting to hard, not soft. SGTM.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5270)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Reviewed 12 of 12 files at r1.
Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---

*[sql/select.go, line 160 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5270#-KCwKZLcwDAJLXVcVHvV:-KCwKZLcwDAJLXVcVHvW:1049310168) ([raw file](https://github.com/cockroachdb/cockroach/blob/d21582821e6360810f4b5778f82c61ea5fc8fcd7/sql/select.go#L160)):*
ignorable nit: `SetLimitHint(soft || s.filter != nil)`

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5270)*
<!-- Sent from Reviewable.io -->

petermattis commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5270#-:-KCzd41TrlQzENCom2Rj:1417088721)*
<!-- Sent from Reviewable.io -->

RaduBerinde committed:
sql: plumb a soft/hard flag for the limit hint

There are situations where if we know for sure the upper layer only needs X rows
we can be more efficient: Nate is looking at making the sortNode use only O(X)
memory in this case. This change improves the SetLimitHint API by adding a
soft vs hard limit flag. Soft limits are "hints", hard limits are "promises".
